mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala

```diff
@@ -421,6 +421,11 @@ class RowMatrix @Since("1.0.0") (
     }
   }
 
+  // The matrix is sparse, if all the rows has sparsity more than 0.5.
+  private def isSparseMatrix: Boolean = {
+    rows.filter(row => row.sparsity() < 0.5).isEmpty()
```

Comment:
I think this could be fine but why 0.5? do we not still have similar problems for less-sparse sparse input? I feel like this isn't fixing the underlying problem - does anyone know why PCA is giving different results for sparse input?

Reply:
Hi @srowen , The underlying assumptions behind the PR are following,
1. For Sparse data, both the methods gives accurate results, but for dense data `computeDenseVectorCovariance` gives more accurate results.
2. `computeDenseVectorCovariance` is costlier than `computeSparseVectorCovariance`.

So based on the above points, we need to restrict the computeSparseVectorCovariance only to sparse matrix.
But how to find the sparse matrix is the problem. In this PR, if more than 50% of the matrix/vector values are sparse, then consider that matrix as sparse.

The 0.5 factor can change to 0.6 or 0.75 to increase the sparsity.

Reply:
Do you have insight into why the dense computation is more accurate? it shouldn't be in theory. I am OK with the change if it's a band-aid that improves results, but not quite sure why this is happening.

Reply:
@KyleLi1985 has done experiments with Sparse and dense data https://github.com/apache/spark/pull/23126#issuecomment-441658381 in the PR: https://github.com/apache/spark/pull/23126. 

The conclusion was for dense data `computeSparseVectorCovariance` doesn't give good accuracy. That is why we have moved to `computeDenseVectorCovariance`. In this PR, I am trying to classify which one to choose dense and sparse path. 
 I can do more experimentation on sparse and dense datasets to verify the accuracy part.

Reply:
>Do you have insight into why the dense computation is more accurate? it shouldn't be in theory.

Yes, For dense case we are using the actual formula to compute the covariance (which is costly). For sparse case we do approximation.